core/src/main/scala/Diet.scala

```diff
@@ -0,0 +1,321 @@
+
+import dogs.Order._
+import dogs.Predef._
+import dogs._
+
+import scala.annotation.tailrec
+
+
+
+/**
+ * Represent discrete operations that can be performed on A
+ */
+trait Discrete[A] extends Order[A] {
+
+  /**
+   * return the successor of x
+   */
+  def succ(x: A): A
+
+  /**
+   * return the predecessor of
+   */
+  def pred(x: A): A
+
+  /**
+   * verify if x and y are consecutive
+   */
+  def adj(x: A, y: A): Boolean = succ(x) == y
+}
+
+/**
+ * Represent a range [x, y] that can be generated using generate() or reverse().
+ */
+trait ARange[A] {
+  def apply(start: A, end: A): ARange[A]
+  def generate()(implicit discrete: Discrete[A]): List[A]
+  def reverse()(implicit discrete: Discrete[A]): List[A]
+}
+
+/**
+ * Implementation of ARange that uses Discrete[A] to generate items in the range
+ */
+sealed class DRange[A] private [dogs] (val start: A, val end: A) extends ARange[A] {
+  override def generate()(implicit discrete: Discrete[A]): List[A] = {
+    @tailrec def genRange(a: A, b: A, xs: List[A])(implicit discrete: Discrete[A]): List[A] = {
+      if (discrete.compare(a, b) == EQ) {
+        xs ::: Nel(a, El())
+      } else if (discrete.adj(a, b)) {
+        xs ::: Nel(a, Nel(b, El()))
+      }
+      else {
+        genRange(discrete.succ(a), b, xs ::: (Nel(a, El())))
+      }
+    }
+
+    genRange(start, end, El())
+  }
+
+  override def reverse()(implicit discrete: Discrete[A]): List[A] = {
+    @tailrec def genRev(a: A, b: A, xs: List[A])(implicit discrete: Discrete[A]): List[A] = {
+      if (discrete.compare(a, b) == EQ) {
+        xs ::: Nel(a, El())
+      } else if (discrete.adj(a, b)) {
+        xs ::: Nel(a, Nel(b, El()))
+      }
+      else {
+        genRev(discrete.pred(a), b, xs ::: (Nel(a, El())))
+      }
+    }
+
+    genRev(end, start, El())
+  }
+
+  override def apply(start: A, end: A): DRange[A] = DRange.apply(start, end)
+}
+
+object DRange {
+  def apply[A](x: A, y: A) = new DRange[A](x, y)
+}
+
+/**
+ * Discrete Interval Encoding Tree (Diet).
+ * It stores subsets of types having a total order, a predecessor and a successor function described by Discrete[A]
+ *
+ * Diet is a binary search tree where each node contains a range of values and the set of all nodes is a set of
+ * disjoint sets.
+ *
+ * In the best case, when there are no "holes" in the stored set, the interval representation consists of just one
+ * single interval (node) and finding, inserting and deleting operations are O(1). In the worse case, where there
+ * are not two adjacent elements in the set, the representation is equivalent to a binary search tree.
+ */
+sealed abstract class Diet[A] {
+  import Diet._
+
+  val isEmpty: Boolean
+
+  /**
+   * verify x is a value is in the tree
+   */
+  def contains(x: A)(implicit discrete: Discrete[A]): Boolean = this match {
+    case EmptyDiet()          =>  false
+    case DietNode(a,b, l, r)  =>  (discrete.compare(x, a), discrete.compare(x, b)) match {
+      case (EQ, _)  =>  true
```

Comment:
I think we should just add "ge" and "le" methods to order to be able to ask about "greater than or equals" or "less than or equals"